### PR TITLE
refac: implement linear combination

### PR DESCRIPTION
**Pull request by chokobole**
# Description

This replaces `Polynomial`'s `Linearize(InPlace)()` methods and manual linear combinations.
    
This method covers all 3 cases.
    
- AdditiveSemigroup's scalar multiplication e.g, affine point
- Ring's multiplication and addition e.g, polynomial
- Field's multiplication and addition e.g, prime field

### CODE REVIEW

**Comment by dongchangYoo** [comment]
In general, "Linear Combination" means inner product of two vectors as following link; https://en.wikipedia.org/wiki/Linear_combination.

**Review by Insun35 — APPROVED** [review]
LGTM

**Comment by dongchangYoo** [comment]
In all citation papers of the Plonk paper, the term `random linear combination` was used as a common noun, referring to the combination with a random vector as a `random linear combination`.

There are two recommandations.
- use 'LinearCombinationWithRandomPowers" instead of "LinearCombination" (Is it too long name?)
- overloading functions
  - `F LinearCombination(const Container1<T1>& vec1, const Contatiner<T2>& vec2)` for calculate inner product
  - `F LinearCombination(const Contatiner1<T>& vec, const Point& r)` -> F for calculate `Sum(vec[i] * r^i)`

File: tachyon/math/base/semigroups.h [hunk 1]
```diff
@@ -380,6 +381,76 @@ class AdditiveSemigroup {
...
 
+  // Linear combination
+  // - forward: a₀ * rⁿ⁻¹ + a₁ * rⁿ⁻² + ... + aₙ₋₁
+  // - backward: a₀ + a₁ * r + ... + aₙ₋₁ * rⁿ⁻¹
+  // NOTE(chokobole): For a performance reason, I would recommend the
+  // |LinearCombinationInPlace()| if possible.
+  template <bool Forward, typename Container, typename T,
+            typename ReturnTy =
+                typename internal::AdditiveSemigroupTraits<G>::ReturnTy>
+  constexpr static ReturnTy LinearCombination(const Container& values,
```
Comment (TomTaehoonKim) [inline tachyon/math/base/semigroups.h:392]: Unlike LinearCombinationInPlace, did you intend to return Zero when size of values is 0?

**Review by TomTaehoonKim — CHANGES_REQUESTED** [review]

**Comment by chokobole** [comment]
> In all citation papers of the Plonk paper, the term `random linear combination` was used as a common noun, referring to the combination with a random vector as a `random linear combination`.
> 
> There are two recommandations.
> 
> * use 'LinearCombinationWithRandomPowers" instead of "LinearCombination" (Is it too long name?)
> * overloading functions
>   
>   * `F LinearCombination(const Container1<T1>& vec1, const Contatiner<T2>& vec2)` for calculate inner product
>   * `F LinearCombination(const Contatiner1<T>& vec, const Point& r)` -> F for calculate `Sum(vec[i] * r^i)`

Can't we just think that `RandomLinearCombination` without `Random`? I have to think more about alternative suggestion because we aren't working on linear algebra, and we refer to it with `InnerProduct` instead mostly when considering `IPA`?

File: tachyon/math/base/semigroups.h (hunk 1 above)
Comment (chokobole) [reply tachyon/math/base/semigroups.h:392]: Yes, I believe `LinearCombination()` should do like that and `LinearCombinationInPlace()` can't return anything if empty arrays are given.

**Comment by dongchangYoo** [comment]
> > In all citation papers of the Plonk paper, the term `random linear combination` was used as a common noun, referring to the combination with a random vector as a `random linear combination`.
> > There are two recommandations.
> > 
> > * use 'LinearCombinationWithRandomPowers" instead of "LinearCombination" (Is it too long name?)
> > * overloading functions
> >   
> >   * `F LinearCombination(const Container1<T1>& vec1, const Contatiner<T2>& vec2)` for calculate inner product
> >   * `F LinearCombination(const Contatiner1<T>& vec, const Point& r)` -> F for calculate `Sum(vec[i] * r^i)`
> 
> For the first recommendation, it seems long to me as well. For the second recommendation, I have to think more about alternative suggestions because we aren't working on linear algebra, and we refer to it with `InnerProduct` instead mostly when considering `IPA`?
> 
> Can't we just think that `RandomLinearCombination` without `Random`? For me, it doesn't sound that strange!

This probably means that they you keep the current function name as it is, right?
Understood. This conversation thread has been resolved.

**Review by dongchangYoo — APPROVED** [review]
LGTM

**Review by TomTaehoonKim — APPROVED** [review]
LGTM

**Pull request merged**